lesson-15/src/index.js

```diff
@@ -0,0 +1,140 @@
+        id: product.id,
+        title: product.title
+      }))
+    } return `категория ${nameOfCategory} не найдена`
+  }
+
+  getCategoryProducts(listOfCategories, 'Electronics').then((result) => {
+    console.log(result)
+  })
+
+  getCategoryProducts(listOfCategories, 'Groceries').then((result) => {
+    console.log(result)
+  })
+
+  // 5
+  async function fetchProductsNamesAndPrises(arrayOfCategories, nameOfCategory) {
+    const category = arrayOfCategories.find((categoryItem) => categoryItem.name === nameOfCategory)
+    if (category !== undefined) {
+      const response = await fetch(category.url)
+      const listOfProducts = await response.json()
+      return Object.fromEntries(listOfProducts.products.map((product) => [product.title, product.price]))
+    } return `категория ${nameOfCategory} не найдена`
+  }
+
+  fetchProductsNamesAndPrises(listOfCategories, 'Electronics').then((result) => {
+    console.log(result)
+  })
+
+  fetchProductsNamesAndPrises(listOfCategories, 'Groceries').then((result) => {
+    console.log(result)
+  })
+
+  // 6
+
+  async function getAndGroupsProductByCategory() {
+    const response = await fetch('https://dummyjson.com/products?limit=20&select=id,title,category')
+    const data = await response.json()
+    const groupedByCategory = data.products.reduce((acc, product) => {
+      if (!acc[product.category]) {
+        acc[product.category] = []
+      }
+      acc[product.category].push({
+        id: product.id,
+        title: product.title
+      })
+      return acc
+    }, {})
+    console.log(groupedByCategory)
+  }
+  getAndGroupsProductByCategory()
+
+  // 7
+  async function fetchListOFProductsCountByCategories(category) {
+    const response = await fetch(`https://dummyjson.com/products/category/${category}`)
+    const listOfProducts = await response.json()
+    return listOfProducts.products.length
+  }
+
+  async function getAarrayOfCategories() {
+    const response = await fetch('https://dummyjson.com/products?limit=20&select=title,category')
+    const data = await response.json()
+    const arrayOfCategories = data.products.reduce((acc, product) => {
+      if (acc.includes(product.category)) {
+        return acc
+      } return [...acc, product.category]
+    }, [])
+    return arrayOfCategories
+  }
+
+  async function getProductByCategoryAndRerurnQuantity() {
+    const arrayOfCategories = await getAarrayOfCategories()
+    console.log(arrayOfCategories)
+    const productGroupsCountFetchArray = arrayOfCategories.map((category) => fetchListOFProductsCountByCategories(category))
+    console.log(productGroupsCountFetchArray)
+    const productsInSameGropsCount = await Promise.all(productGroupsCountFetchArray)
+    console.log('same products categories count', productsInSameGropsCount)
+    console.log(Object.fromEntries(
+      arrayOfCategories.map((tag, index) => [tag, productsInSameGropsCount[index]])
+    ))
+  }
+  getProductByCategoryAndRerurnQuantity()
```

Comment:
```suggestion
  async function getCategoryProductsCount(category) {
    const response = await fetch(`https://dummyjson.com/products/category/${category}`)
    const { products } = await response.json()

    return products.length
  }

  async function getProductCategories() {
    const response = await fetch('https://dummyjson.com/products?limit=20&select=title,category')
    const { products } = await response.json()

    return products.reduce((acc, { category }) => (
      acc.includes(category) ? acc : [...acc, category]
    ), [])
  }

  async function getProductCategoriesProductsCount() {
    const categories = await getProductCategories()
    console.log(categories)
    const productsInSameGroupsCount = await Promise.all(categories.map((category) => (
      getCategoryProductsCount(category)
    )))
    console.log('same products categories count', productsInSameGroupsCount)

    return Object.fromEntries(
      categories.map((category, index) => [category, productsInSameGroupsCount[index]])
    )
  }

  console.log(await getProductCategoriesProductsCount())
```

зарефакторил по заметкам из предыдущих заданий